### PR TITLE
Add an image text task, datamodule, basic model

### PR DESCRIPTION
**Pull request by teexone**
Add an image to text task and update ~innofw.utils.defaults with default datamodule class for this task

Add the `ImageToTextDataModule` for the image to text task. That is a PyTorch Lightning datamodule that
operates with folders that has Images folder and
captions.txt file wil comma-separated mapping between an image and a caption.

Add the `BasicImageToTextEncoderDecoder` model which has an encoder convolutional neural net (commonly image feature extractor) and a decoder recurrent neural net (LSTM)

Add an experiment to train the abovementioned model on [Flickr8 dataset](https://www.kaggle.com/datasets/adityajn105/flickr8k)

### CODE REVIEW

File: innofw/core/datamodules/lightning_datamodules/image_to_text.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        return DataLoader(
+            ImageToTextDatamodule.ImageToTextDataset(
+                self.train_source,
+                self.train_df,
+                self.word2int,
+                self.max_caption_length
+            )
+        )
+    
+    def predict_dataloader(self):
```
Comment (QazyBi) [inline innofw/core/datamodules/lightning_datamodules/image_to_text.py:112]: does this function have any meaning?

File: innofw/core/datamodules/lightning_datamodules/image_to_text.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+        return super().setup_train_test_val()
+    
+    def setup_infer(self):
+        return super().setup_infer()
+
+    def save_preds(self, preds, stage: Stages, dst_path: Path):
+        return super().save_preds(preds, stage, dst_path)
+    
+    def test_dataloader(self):
```
Comment (QazyBi) [inline innofw/core/datamodules/lightning_datamodules/image_to_text.py:99]: does this function have any meaning?

File: innofw/core/datamodules/lightning_datamodules/image_to_text.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+        if self.word2int is None:
+            self.word2int = dict({'<pad>': 1, '<s>': 2, '</s>': 3, '<?>': 0})
+            word_set = set()
+            for sentence in self.train_df['caption']:
+                word_set.update(ImageToTextDatamodule.split_sentence(sentence))
+            self.word2int.update(enumerate(word_set, start=len(self.word2int)))
+
+        return super().setup_train_test_val()
+    
+    def setup_infer(self):
```
Comment (QazyBi) [inline innofw/core/datamodules/lightning_datamodules/image_to_text.py:93]: does this function have any meaning?

File: innofw/core/datamodules/lightning_datamodules/image_to_text.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        return super().setup_infer()
+
+    def save_preds(self, preds, stage: Stages, dst_path: Path):
```
Comment (QazyBi) [inline innofw/core/datamodules/lightning_datamodules/image_to_text.py:96]: does this function have any meaning?

File: innofw/core/datamodules/lightning_datamodules/image_to_text.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+                self.train_source,
+                self.train_df,
+                self.word2int,
+                self.max_caption_length
+            )
+        )
+    
+    def predict_dataloader(self):
+        return super().predict_dataloader()
+
```
Comment (QazyBi) [inline innofw/core/datamodules/lightning_datamodules/image_to_text.py:114]: I don't see val_dataloader, where it is?

File: innofw/core/datamodules/lightning_datamodules/image_to_text.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+        sentence = ImageToTextDatamodule.split_sentence(sentence)
+        sentence = [words.get(x, words.get('<?>')) for x in sentence]
+        if len(sentence) + 2 > length:
+            raise ValueError(f"Unable to encode sentence of size {len(sentence)}\
+                             in a sentence of target length {length}")
+        sentence = [words.get("<s>")] + sentence + [words.get('</s>')]
+        while len(sentence) < length: sentence.append(words.get('<pad>'))
+        return sentence
+
+    class ImageToTextDataset(Dataset):
```
Comment (QazyBi) [inline innofw/core/datamodules/lightning_datamodules/image_to_text.py:32]: Why not to separate this class to a file and add methods `split_text` and `encode_sentence` to it?

File: innofw/core/models/torch/architectures/image_to_text/basic_encoder_decoder.py [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+                average_pool_size=encoder_avg_pool_size,
+            )
+
+           
+            self.lstm = torch.nn.LSTMCell(
+                input_size=encoder_avg_pool_size ** 2 * encoder_network_size * decoder_hidden_size,
+                hidden_size=decoder_hidden_size
+            )
+
+            self.device = torch.device('cpu')
```
Comment (QazyBi) [inline innofw/core/models/torch/architectures/image_to_text/basic_encoder_decoder.py:77]: apparently this is an old code that has no use right now

File: innofw/core/models/torch/architectures/image_to_text/basic_encoder_decoder.py [hunk 8]
```diff
@@ -0,0 +1,145 @@
+import torch
+from typing import Dict, Union
+
+class BasicImageToTextEncoderDecoder(torch.nn.Module):
+    class Encoder(torch.nn.Module):
```
Comment (QazyBi) [inline innofw/core/models/torch/architectures/image_to_text/basic_encoder_decoder.py:5]: Can you add a paper on which this work is based on?

File: innofw/core/models/torch/lightning_modules/image_to_text.py [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+                scheduler_cfg, 
+                max_caption_length: int=128, 
+                use_teacher_forcing=False,
+                *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self.model = model
+        self.max_caption_length = max_caption_length
+        self.use_teacher_forcing = use_teacher_forcing
+        self.criterion = torch.nn.CrossEntropyLoss()
```
Comment (QazyBi) [inline innofw/core/models/torch/lightning_modules/image_to_text.py:18]: losses parameter is not used

File: innofw/core/models/torch/lightning_modules/image_to_text.py [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+    def configure_optimizers(self):
+        return torch.optim.Adam(self.model.parameters(), lr=1E-5)
```
Comment (QazyBi) [inline innofw/core/models/torch/lightning_modules/image_to_text.py:21]: optimizers_cfg parameter is not used

Comment (QazyBi) [reply innofw/core/models/torch/lightning_modules/image_to_text.py:21]: no scheduler?

File: innofw/core/models/torch/lightning_modules/image_to_text.py [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+    
+    def setup(self, stage: str):
+        if self.trainer and self.trainer.datamodule and not self.model.is_ready:
+            self.model.initialize(self.trainer.datamodule.word2int)
+
+    def training_step(self, batch, batch_ids):
+        self.model.device = self.device
```
Comment (QazyBi) [inline innofw/core/models/torch/lightning_modules/image_to_text.py:28]: this line also seems redundant now

**Review by QazyBi — CHANGES_REQUESTED** [review]
